File: Python/algorithmic_toolbox/11_fractional_knapsack.py
import sys
from collections import namedtuple

Item = namedtuple("Item", ["value", "weight"])


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_items(lines, nr_items):
    items = []
    ix = 0
    while ix < nr_items:
        items.append(Item(*convert_to_intlist(lines[ix])))
        ix += 1
    return items


def fractional_knapsack(items, capacity):
    sorted_items = sorted(items, key=lambda x: x.value / x.weight, reverse=True)
    total_value = 0
    for value, weight in sorted_items:
        if capacity == 0:
            return total_value
        amount = min(weight, capacity)
        total_value += amount * (value / weight)
        capacity -= amount
    return total_value


def main():
    data = sys.stdin.read().splitlines()
    nr_items, capacity = convert_to_intlist(data[0])
    items = read_items(data[1:], nr_items)
    value = fractional_knapsack(items, capacity)
    print(value)

if __name__ == "__main__":
    main()

File: Julia/algorithmic_toolbox/11-fractional-knapsack.jl
converttointlist(line) = map(x -> parse(Int, x), split(line))


function readitems(nritems)
    items = []
    for _ in 1:nritems
        value, weight = readline() |> converttointlist
        push!(items, (value=value, weight=weight))
    end
    items
end


function fractionalknapsack(items, capacity)
    sorteditems = sort(items; by=x -> x.value / x.weight, rev=true)
    totalvalue = 0
    for (value, weight) in sorteditems
        if capacity == 0
            return totalvalue
        end
        amount = min(weight, capacity)
        totalvalue += amount * (value / weight)
        capacity -= amount
    end
    totalvalue
end


function main()
    nritems, capacity = readline() |> converttointlist
    items = readitems(nritems)
    value = fractionalknapsack(items, capacity)
    println(value)
end


main()
